src/Neo/Ledger/SmartThrottler.cs

```diff
@@ -0,0 +1,185 @@
+    /// <param name="system">The Neo system</param>
+    public SmartThrottler(MemoryPool memoryPool, NeoSystem system)
+    {
+        _memoryPool = memoryPool;
+        _system = system;
+        _maxTransactionsPerSecond = (uint)system.Settings.MemPoolSettings.MaxTransactionsPerSecond;
+        _lastResetTime = TimeProvider.Current.UtcNow;
+        _lastBlockTimestamp = _lastResetTime.ToTimestampMS();
+        _averageFee = CalculateAverageFee(null);
+    }
+
+    /// <summary>
+    /// Determines whether a new transaction should be accepted
+    /// </summary>
+    /// <param name="tx">The transaction to be evaluated</param>
+    /// <returns>True if the transaction should be accepted, false otherwise</returns>
+    public bool ShouldAcceptTransaction(Transaction tx)
+    {
+        var now = TimeProvider.Current.UtcNow;
+        if (now - _lastResetTime >= TimeSpan.FromSeconds(1))
+        {
+            _transactionsThisSecond = 0;
+            _lastResetTime = now;
+            AdjustThrottling(null);
+            _averageFee = CalculateAverageFee(null);
+        }
+
+        // Check if we've hit the tx limit and it's not high priority
+        if ((_transactionsThisSecond >= _maxTransactionsPerSecond) && !IsHighPriorityTransaction(tx))
+            return false;
+
+        _transactionsThisSecond++;
+        return true;
+    }
+
+    /// <summary>
+    /// Updates the network state after a new block is added
+    /// </summary>
+    /// <param name="block">The newly added block</param>
+    public void UpdateNetworkState(Block block)
+    {
+        var currentTime = TimeProvider.Current.UtcNow.ToTimestampMS();
+        var blockTime = currentTime - _lastBlockTimestamp;
+
+        _recentBlockTimes.Enqueue(blockTime);
+        if (_recentBlockTimes.Count > BlockTimeWindowSize)
+            _recentBlockTimes.Dequeue();
+
+        _lastBlockTimestamp = currentTime;
+        _unconfirmedTxCount = _memoryPool.Count;
+        _averageFee = CalculateAverageFee(block);
+
+        AdjustThrottling(block);
+    }
+
+    /// <summary>
+    /// Adjusts throttling parameters based on current network conditions
+    /// </summary>
+    private void AdjustThrottling(Block? block)
+    {
+        var memoryPoolUtilization = (double)_memoryPool.Count / _system.Settings.MemoryPoolMaxTransactions;
+        var networkLoad = EstimateNetworkLoad(block);
+
+        _maxTransactionsPerSecond = CalculateOptimalTps(memoryPoolUtilization, networkLoad, block);
+    }
+
+    /// <summary>
+    /// Estimates current network load
+    /// </summary>
+    /// <returns>An integer between 0 and 100 representing the estimated network load</returns>
+    private int EstimateNetworkLoad(Block? block)
+    {
+        var load = 0;
+
+        // 1. Memory pool utilization (30% weight)
+        var memPoolUtilization = (double)_memoryPool.Count / _system.Settings.MemoryPoolMaxTransactions;
```

Comment:
get this value from previous line before calling the method, it  is called just once as it is right now